python/pyarrow/table.pxi

```diff
@@ -2789,13 +2798,36 @@ cdef class RecordBatch(_Tabular):
         ----
         n: [2,4,5,100]
         name: ["Flamingo","Horse","Brittle stars","Centipede"]
+        >>> new_names = {"n_legs": "n", "animals": "name"}
+        >>> batch.rename_columns(new_names)
+        pyarrow.RecordBatch
+        n: int64
+        name: string
+        ----
+        n: [2,4,5,100]
+        name: ["Flamingo","Horse","Brittle stars","Centipede"]
         """
         cdef:
             shared_ptr[CRecordBatch] c_batch
             vector[c_string] c_names
 
-        for name in names:
-            c_names.push_back(tobytes(name))
+        if isinstance(names, list):
+            for name in names:
+                c_names.push_back(tobytes(name))
+        else:
```

Comment:
Since you added an `isinstance` check, I would also add a type check for `dict`, and output a `TypeError` if neither list nor dict are passed in. Otherwise, looks good!

Reply:
Implemented in https://github.com/apache/arrow/pull/40645/commits/b0973586148856daa8e961e9c2826c8109cd9d05